### DISCUSSION TITLE
removing an entity from memory within a foreach loop

### DISCUSSION THREAD

**Discussion by ghost** [stride3d/stride#1348] [Q&A]
foreach(var entity in Entity.Scene.Entities)
            {

                if(entity.Name == "Cube")
                {
                    count++;
                    if(entity.Transform.Position.Y < -100)
                    {
                        Entity.Scene.Entities.Remove(entity);
                        
                    }
                }
            }

how do i remove the entity from memory after removing it from the scene? "entity = null;" doesn't work as it's a read only variable.

**Comment by tebjan** [comment]
> how do i remove the entity from memory after removing it from the scene?

Are you coming from a C or C++ environment? Since C# is a managed language, all objects that have no reference to it will be collected by the garbage collector. See: [Automatic Memory Management](https://docs.microsoft.com/en-us/dotnet/standard/automatic-memory-management)

You only need to clean up objects that have a `Dispose()` method, this usually cleans up unmanaged memory or other system resources. See [Using objects that implement IDisposable](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/using-objects)

**Reply by ghost** [reply]
neither. i'm learning C# with no prior knowledge of any C language

**Reply by ghost** [reply]
i'm removing the objects from the scene, but they're staying in memory, even after calling entity.Dispose()

**Reply by tebjan** [reply]
then there is another reference to them somewhere. i think you should first get familiar with general C# coding. there are many good tutorials on youtube.

Did you also go through these tutorials? https://doc.stride3d.net/4.0/en/tutorials/csharpbeginner/removing-entities.html

**Reply by ghost** [reply]
i have gone through that, yes

**Reply by ghost** [reply]
it says to set the variable for the object to null, but that doesn't work in a for loop. so i'm looking for another way to do it.

**Reply by Eideren** [reply]
You should really take a look into how reference types, enumerators and the garbage collector work before continuing otherwise you'll just fall into similar issues and become frustrated.
Here's how you remove entities from the scene, it will be removed from memory if nothing else is referencing them and the GC deemed it necessary to clean up memory.
```cs
for(int i = Entity.Scene.Entities.Count-1; i >= 0; i--)
{
    var entity = Entity.Scene.Entities[i];
    if(entity.Name == "Cube" && entity.Transform.Position.Y < -100)
    {
        Entity.Scene.Entities.RemoveAt(i);
    }
}
```
If you are confused or have further questions about this topic, have a look at the many resources the internet has on the things I listed above, they will be answer your questions in far more detail than we ever could.

**Reply by ghost** [reply]
public override void Update()
        {
            int count = 0;
            foreach(var entity in Entity.Scene.Entities)
            {

                if(entity.Name == "Cube")
                {
                    count++;
                    if(entity.Transform.Position.Y < -100)
                    {
                        Entity.Scene.Entities.Remove(entity);
                    }
                }
            }
            if(count < 1000)
            {
                for(int i = 0; i < 3; i++)
                {
                    var box = Content.Load<Prefab>("Box").Instantiate();
                    foreach(var entity in box)
                    {
                        entity.Transform.Position.Y = (float)(entity.Transform.Position.Y + 5);
                        Entity.Scene.Entities.Add(entity);
                    }
                }
            }
        }
here's all the code. once an object is removed from the scene tree, i can't access it anymore, and it shouldn't be referenced anywhere else either, so i don't know why it isn't doing garbage collection on them. if i leave the project running for a while it literally uses all my remaining RAM, which is something i don't think it should be doing.

**Reply by Eideren** [reply]
Your entity is probably named `Box`, not `Cube`, given the URL you use so it'll never reach 1000 since there are no `Cube` in your scene, just a bunch of `Box`. Also removing from a collection being enumerated and continuing with the iteration is a runtime error, again, read up on enumerators.
This issue is unrelated to stride, your logic is just not sound. Use your IDE's debugger and step through the code, to have a better idea of what's going on.

**Reply by ghost** [reply]
Box is a prefab, containing 3 Cubes. it loads an instance of Box, and then adds each Cube from the prefab to the scene. it gets to 1000 Cubes, then stops. this is correct behaviour. when an object is removed, the count decreases below 1000. this is intended. However i do see your point that removing something from the scene while looping through the scene may cause issues regarding references, therefore preventing the GC from removing the objects after unloading them from the scene.
